File: 连通块中点的数量.py
class DSU: # 并查集
    def __init__(self, n):
        self.pre = [i for i in range(n + 1)]
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        if self.pre[x] == x:
            return x
        self.pre[x] = self.find(self.pre[x])
        return self.pre[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]
    
n, m = map(int, input().split())
dsu = DSU(n)
for _ in range(m):
    op = list(map(str, input().split()))
    if op[0] == "C":
        dsu.merge(int(op[1]), int(op[2]))
    elif op[0] == "Q1":
        print("Yes" if dsu.same(int(op[1]), int(op[2])) else "No")
    else:
        print(dsu.size(int(op[1])))

File: 连通块中点的数量.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n);
        std::iota(pre.begin(), pre.end(), 0);
        siz.assign(n, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, m;
    cin >> n >> m;
    DSU dsu(n + 1);
    while (m--) {
        string op;
        int a, b;
        cin >> op >> a;
        if (op == "C") {
            cin >> b;
            dsu.merge(a, b);
        } else if (op == "Q1") {
            cin >> b;
            cout << (dsu.same(a, b) ? "Yes" : "No") << endl; 
        } else {
            cout << dsu.size(a) << endl;
        }
    }
    return 0;
}
